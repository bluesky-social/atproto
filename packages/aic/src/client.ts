import { DidableKey } from '@adxp/crypto'
import axios from 'axios'
import { Document, hashAndFindDid } from './document'
import { Create, create, Operation, UpdateOperation } from './types'
import * as cbor from '@ipld/dag-cbor'
import * as uint8arrays from 'uint8arrays'
import { check, cidForData } from '@adxp/common'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<Document> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op: Partial<Create> = {
      type: 'create',
      signingKey: signingKey.did(),
      recoveryKey,
      username,
      service,
      prev: null,
    }
    const signedOp = await this.signOperation(op, signingKey)
    if (!check.is(signedOp, create)) {
      throw new Error('Not a valid create operation')
    }
    const did = await hashAndFindDid(signedOp)
    await axios.post(`${this.url}/${did}`, signedOp)
    return did
  }

  async makeUpdate(
    did: string,
    op: Partial<UpdateOperation>,
    signingKey: DidableKey,
  ) {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    const prev = await cidForData(log[log.length - 1])
    const signedOp = this.signOperation(
      {
        ...op,
        prev: prev.toString(),
      },
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, signedOp)
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_signing_key', key: newKey },
      signingKey,
    )
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_recovery_key', key: newKey },
      signingKey,
    )
  }

  async udpateUsername(did: string, username: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'update_username', username },
      signingKey,
    )
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    return this.makeUpdate(did, { type: 'update_service', service }, signingKey)
  }
}

export default AicClient
