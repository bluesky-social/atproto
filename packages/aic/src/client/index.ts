import axios from 'axios'
import { CID } from 'multiformats/cid'
import { DidableKey } from '@adxp/crypto'
import { check, cidForData } from '@adxp/common'
import * as operations from '../lib/operations'
import * as t from '../lib/types'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<t.DocumentData> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<t.Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op = await operations.create(
      signingKey,
      recoveryKey,
      username,
      service,
    )
    if (!check.is(op, t.def.createOp)) {
      throw new Error('Not a valid create operation')
    }
    const did = await operations.didForCreateOp(op)
    await axios.post(`${this.url}/${did}`, op)
    return did
  }

  async getPrev(did): Promise<CID> {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    return cidForData(log[log.length - 1])
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateSigningKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateRecoveryKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateUsername(did: string, username: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateUsername(
      username,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateService(
      service,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }
}

export default AicClient
