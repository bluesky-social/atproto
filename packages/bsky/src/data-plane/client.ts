import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import * as ui8 from 'uint8arrays'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { getDidKeyFromMultibase } from '@atproto/identity'
import { Service } from '../proto/bsky_connect'

export type DataPlaneClient = PromiseClient<typeof Service>
type BaseClient = { lib: DataPlaneClient; url: URL }
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  baseUrls: string[],
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = baseUrls.map((baseUrl) => createBaseClient(baseUrl, opts))
  assert(clients.length > 0, 'no clients available')
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client.lib[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): BaseClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return {
    lib: createPromiseClient(Service, transport),
    url: new URL(baseUrl),
  }
}

const getRemainingClients = (clients: BaseClient[], lastClient: BaseClient) => {
  if (clients.length < 2) return clients // no clients to choose from
  if (lastClient.url.port) {
    // if the last client had a port, we attempt to exclude its whole host.
    const maybeRemaining = clients.filter(
      (c) => c.url.hostname !== lastClient.url.hostname,
    )
    if (maybeRemaining.length) {
      return maybeRemaining
    }
  }
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
