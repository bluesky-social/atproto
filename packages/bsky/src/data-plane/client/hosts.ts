import { Etcd3 } from 'etcd3'
import { EtcdMap } from '../../etcd'
import { dataplaneLogger as logger } from '../../logger'

/**
 * Interface for a reactive list of hosts, i.e. for use with the dataplane client.
 */
export interface HostList {
  get: () => Iterable<string>
  onUpdate(handler: HostListHandler): void
}

type HostListHandler = (hosts: Iterable<string>) => void

/**
 * Maintains a reactive HostList based on a simple setter.
 */
export class BasicHostList implements HostList {
  private hosts: Iterable<string>
  private handlers: HostListHandler[] = []

  constructor(hosts: Iterable<string>) {
    this.hosts = hosts
  }

  get() {
    return this.hosts
  }

  set(hosts: Iterable<string>) {
    this.hosts = hosts
    this.update()
  }

  private update() {
    for (const handler of this.handlers) {
      handler(this.hosts)
    }
  }

  onUpdate(handler: HostListHandler) {
    this.handlers.push(handler)
  }
}

/**
 * Maintains a reactive HostList based on etcd key values under a given key prefix.
 * When fallback is provided, ensures that this fallback is used whenever no hosts are available.
 */
export class EtcdHostList implements HostList {
  private kv: EtcdMap
  private inner = new BasicHostList(new Set())
  private fallback: Set<string>

  constructor(etcd: Etcd3, prefix: string, fallback?: string[]) {
    this.fallback = new Set(fallback)
    this.kv = new EtcdMap(etcd, prefix)
    this.update() // init fallback if necessary
    this.kv.watcher.on('connected', (res) => {
      logger.warn(
        { watcherId: this.kv.watcher.id, header: res.header },
        'EtcdHostList connected',
      )
    })
    this.kv.watcher.on('disconnected', (err) => {
      logger.warn(
        { watcherId: this.kv.watcher.id, err },
        'EtcdHostList disconnected',
      )
    })
    this.kv.watcher.on('error', (err) => {
      logger.error({ watcherId: this.kv.watcher.id, err }, 'EtcdHostList error')
    })
  }

  async connect() {
    await this.kv.connect()
    this.update()
    this.kv.onUpdate(() => this.update())
  }

  get() {
    return this.inner.get()
  }

  private update() {
    const hosts = new Set<string>()
    for (const host of this.kv.values()) {
      if (URL.canParse(host)) {
        hosts.add(host)
      }
    }
    if (hosts.size) {
      this.inner.set(hosts)
    } else if (this.fallback.size) {
      this.inner.set(this.fallback)
    }
  }

  onUpdate(handler: HostListHandler) {
    this.inner.onUpdate(handler)
  }
}
