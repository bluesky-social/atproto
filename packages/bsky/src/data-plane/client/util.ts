import { Code, ConnectError } from '@connectrpc/connect'
import * as ui8 from 'uint8arrays'
import { getDidKeyFromMultibase } from '@atproto/identity'

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
