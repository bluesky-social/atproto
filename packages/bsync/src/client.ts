import {
  Interceptor,
  PromiseClient,
  createPromiseClient,
} from '@connectrpc/connect'
import {
  ConnectTransportOptions,
  createConnectTransport,
} from '@connectrpc/connect-node'
import { Service } from './proto/bsync_connect'

export type BsyncClient = PromiseClient<typeof Service>

export const createClient = (opts: ConnectTransportOptions): BsyncClient => {
  const transport = createConnectTransport(opts)
  return createPromiseClient(Service, transport)
}

export const authWithApiKey =
  (apiKey: string): Interceptor =>
  (next) =>
  (req) => {
    req.header.set('authorization', `Bearer ${apiKey}`)
    return next(req)
  }
