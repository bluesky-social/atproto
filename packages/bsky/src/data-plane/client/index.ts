import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { Service } from '../../proto/bsky_connect'
import { HostList } from './hosts'

export * from './hosts'
export * from './util'

export type DataPlaneClient = PromiseClient<typeof Service>
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  hostList: HostList,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = new DataPlaneClients(hostList, opts)
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients.get()
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

/**
 * Uses a reactive HostList in order to maintain a pool of DataPlaneClients.
 * Each DataPlaneClient is cached per host so that it maintains connections
 * and other internal state when the underlying HostList is updated.
 */
class DataPlaneClients {
  private clients: DataPlaneClient[] = []
  private clientsByHost = new Map<string, DataPlaneClient>()

  constructor(
    private hostList: HostList,
    private clientOpts: {
      httpVersion?: HttpVersion
      rejectUnauthorized?: boolean
    },
  ) {
    this.refresh()
    this.hostList.onUpdate(() => this.refresh())
  }

  get(): readonly DataPlaneClient[] {
    return this.clients
  }

  private refresh() {
    this.clients = []
    for (const host of this.hostList.get()) {
      let client = this.clientsByHost.get(host)
      if (!client) {
        client = this.createClient(host)
        this.clientsByHost.set(host, client)
      }
      this.clients.push(client)
    }
  }

  private createClient(host: string) {
    return createBaseClient(host, this.clientOpts)
  }
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): DataPlaneClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return createPromiseClient(Service, transport)
}

const getRemainingClients = (
  clients: readonly DataPlaneClient[],
  lastClient: DataPlaneClient,
) => {
  if (clients.length < 2) return clients // no clients to choose from
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: readonly T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}
