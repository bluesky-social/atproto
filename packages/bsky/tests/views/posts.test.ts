import AtpAgent from '@atproto/api'
import { TestEnvInfo, runTestEnv } from '@atproto/dev-env'
import { forSnapshot, processAll, stripViewerFromPost } from '../_util'
import { SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'

describe('pds posts views', () => {
  let testEnv: TestEnvInfo
  let agent: AtpAgent
  let sc: SeedClient

  beforeAll(async () => {
    testEnv = await runTestEnv({
      dbPostgresSchema: 'bsky_views_posts',
    })
    agent = new AtpAgent({ service: testEnv.bsky.url })
    const pdsAgent = new AtpAgent({ service: testEnv.pds.url })
    sc = new SeedClient(pdsAgent)
    await basicSeed(sc)
    await processAll(testEnv)
  })

  afterAll(async () => {
    await testEnv.close()
  })

  it('fetches posts', async () => {
    const uris = [
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.alice][1].ref.uriStr,
      sc.posts[sc.dids.bob][0].ref.uriStr,
      sc.posts[sc.dids.carol][0].ref.uriStr,
      sc.posts[sc.dids.dan][1].ref.uriStr,
      sc.replies[sc.dids.alice][0].ref.uriStr,
    ]
    const posts = await agent.api.app.bsky.feed.getPosts(
      { uris },
      { headers: sc.getHeaders(sc.dids.alice, true) },
    )

    expect(posts.data.posts.length).toBe(uris.length)
    expect(forSnapshot(posts.data.posts)).toMatchSnapshot()
  })

  it('fetches posts unauthed', async () => {
    const uris = [
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.alice][1].ref.uriStr,
      sc.posts[sc.dids.bob][0].ref.uriStr,
      sc.posts[sc.dids.carol][0].ref.uriStr,
      sc.posts[sc.dids.dan][1].ref.uriStr,
      sc.replies[sc.dids.alice][0].ref.uriStr,
    ]

    const authed = await agent.api.app.bsky.feed.getPosts(
      { uris },
      { headers: sc.getHeaders(sc.dids.alice, true) },
    )
    const unauthed = await agent.api.app.bsky.feed.getPosts({
      uris,
    })
    const stripped = authed.data.posts.map((p) => stripViewerFromPost(p))
    expect(unauthed.data.posts).toEqual(stripped)
  })

  it('handles repeat uris', async () => {
    const uris = [
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.bob][0].ref.uriStr,
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.bob][0].ref.uriStr,
    ]

    const posts = await agent.api.app.bsky.feed.getPosts({ uris })

    expect(posts.data.posts.length).toBe(2)
    const recivedUris = posts.data.posts.map((p) => p.uri).sort()
    const expected = [
      sc.posts[sc.dids.alice][0].ref.uriStr,
      sc.posts[sc.dids.bob][0].ref.uriStr,
    ].sort()
    expect(recivedUris).toEqual(expected)
  })
})
