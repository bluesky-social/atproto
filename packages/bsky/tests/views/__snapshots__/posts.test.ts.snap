// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds posts views fetches posts 1`] = `
Array [
  Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [],
      "viewer": Object {},
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "hey there",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {},
  },
  Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [],
      "viewer": Object {},
    },
    "cid": "cids(2)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 3,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "again",
    },
    "replyCount": 2,
    "repostCount": 1,
    "uri": "record(1)",
    "viewer": Object {},
  },
  Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(3)/cids(1)@jpeg",
      "did": "user(2)",
      "displayName": "bobby",
      "handle": "bob.test",
      "labels": Array [],
      "viewer": Object {
        "followedBy": "record(4)",
        "following": "record(3)",
      },
    },
    "cid": "cids(3)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "bob back at it again!",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(2)",
    "viewer": Object {},
  },
  Object {
    "author": Object {
      "did": "user(4)",
      "handle": "carol.test",
      "labels": Array [],
      "viewer": Object {
        "followedBy": "record(7)",
        "following": "record(6)",
      },
    },
    "cid": "cids(4)",
    "embed": Object {
      "$type": "app.bsky.embed.recordWithMedia#view",
      "media": Object {
        "$type": "app.bsky.embed.images#view",
        "images": Array [
          Object {
            "alt": "tests/image/fixtures/key-landscape-small.jpg",
            "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(5)/cids(5)@jpeg",
            "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(5)/cids(5)@jpeg",
          },
          Object {
            "alt": "tests/image/fixtures/key-alt.jpg",
            "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(5)/cids(6)@jpeg",
            "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(5)/cids(6)@jpeg",
          },
        ],
      },
      "record": Object {
        "record": Object {
          "$type": "app.bsky.embed.record#viewRecord",
          "author": Object {
            "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(3)/cids(1)@jpeg",
            "did": "user(2)",
            "displayName": "bobby",
            "handle": "bob.test",
            "labels": Array [],
            "viewer": Object {
              "followedBy": "record(4)",
              "following": "record(3)",
            },
          },
          "cid": "cids(3)",
          "embeds": Array [],
          "indexedAt": "1970-01-01T00:00:00.000Z",
          "uri": "record(2)",
          "value": Object {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "text": "bob back at it again!",
          },
        },
      },
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 2,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "embed": Object {
        "$type": "app.bsky.embed.recordWithMedia",
        "media": Object {
          "$type": "app.bsky.embed.images",
          "images": Array [
            Object {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "image": Object {
                "$type": "blob",
                "mimeType": "image/jpeg",
                "ref": Object {
                  "$link": "cids(5)",
                },
                "size": 4114,
              },
            },
            Object {
              "alt": "tests/image/fixtures/key-alt.jpg",
              "image": Object {
                "$type": "blob",
                "mimeType": "image/jpeg",
                "ref": Object {
                  "$link": "cids(6)",
                },
                "size": 12736,
              },
            },
          ],
        },
        "record": Object {
          "record": Object {
            "cid": "cids(3)",
            "uri": "record(2)",
          },
        },
      },
      "text": "hi im carol",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(5)",
    "viewer": Object {
      "like": "record(8)",
    },
  },
  Object {
    "author": Object {
      "did": "user(6)",
      "handle": "dan.test",
      "labels": Array [],
      "viewer": Object {
        "following": "record(10)",
      },
    },
    "cid": "cids(7)",
    "embed": Object {
      "$type": "app.bsky.embed.record#view",
      "record": Object {
        "$type": "app.bsky.embed.record#viewRecord",
        "author": Object {
          "did": "user(4)",
          "handle": "carol.test",
          "labels": Array [],
          "viewer": Object {
            "followedBy": "record(7)",
            "following": "record(6)",
          },
        },
        "cid": "cids(4)",
        "embeds": Array [
          Object {
            "$type": "app.bsky.embed.recordWithMedia#view",
            "media": Object {
              "$type": "app.bsky.embed.images#view",
              "images": Array [
                Object {
                  "alt": "tests/image/fixtures/key-landscape-small.jpg",
                  "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(5)/cids(5)@jpeg",
                  "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(5)/cids(5)@jpeg",
                },
                Object {
                  "alt": "tests/image/fixtures/key-alt.jpg",
                  "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(5)/cids(6)@jpeg",
                  "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(5)/cids(6)@jpeg",
                },
              ],
            },
            "record": Object {
              "record": Object {
                "$type": "app.bsky.embed.record#viewRecord",
                "author": Object {
                  "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(3)/cids(1)@jpeg",
                  "did": "user(2)",
                  "displayName": "bobby",
                  "handle": "bob.test",
                  "labels": Array [],
                  "viewer": Object {
                    "followedBy": "record(4)",
                    "following": "record(3)",
                  },
                },
                "cid": "cids(3)",
                "indexedAt": "1970-01-01T00:00:00.000Z",
                "uri": "record(2)",
                "value": Object {
                  "$type": "app.bsky.feed.post",
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "text": "bob back at it again!",
                },
              },
            },
          },
        ],
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "uri": "record(5)",
        "value": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": Object {
            "$type": "app.bsky.embed.recordWithMedia",
            "media": Object {
              "$type": "app.bsky.embed.images",
              "images": Array [
                Object {
                  "alt": "tests/image/fixtures/key-landscape-small.jpg",
                  "image": Object {
                    "$type": "blob",
                    "mimeType": "image/jpeg",
                    "ref": Object {
                      "$link": "cids(5)",
                    },
                    "size": 4114,
                  },
                },
                Object {
                  "alt": "tests/image/fixtures/key-alt.jpg",
                  "image": Object {
                    "$type": "blob",
                    "mimeType": "image/jpeg",
                    "ref": Object {
                      "$link": "cids(6)",
                    },
                    "size": 12736,
                  },
                },
              ],
            },
            "record": Object {
              "record": Object {
                "cid": "cids(3)",
                "uri": "record(2)",
              },
            },
          },
          "text": "hi im carol",
        },
      },
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "embed": Object {
        "$type": "app.bsky.embed.record",
        "record": Object {
          "cid": "cids(4)",
          "uri": "record(5)",
        },
      },
      "facets": Array [
        Object {
          "features": Array [
            Object {
              "$type": "app.bsky.richtext.facet#mention",
              "did": "user(0)",
            },
          ],
          "index": Object {
            "byteEnd": 18,
            "byteStart": 0,
          },
        },
      ],
      "text": "@alice.bluesky.xyz is the best",
    },
    "replyCount": 0,
    "repostCount": 1,
    "uri": "record(9)",
    "viewer": Object {},
  },
  Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [],
      "viewer": Object {},
    },
    "cid": "cids(8)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "reply": Object {
        "parent": Object {
          "cid": "cids(9)",
          "uri": "record(12)",
        },
        "root": Object {
          "cid": "cids(2)",
          "uri": "record(1)",
        },
      },
      "text": "thanks bob",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(11)",
    "viewer": Object {},
  },
]
`;
